### ISSUE TITLE
Parser segfault when parsing CodeCompass

### ISSUE THREAD

**Issue by mcserep** [Ericsson/CodeCompass#471]
With the current state of the `master` branch I tried to parse CodeCompass itself (the same state of the master branch).

I get the following segfault:

> [INFO] (21/138) Parsing /home/mate/cc/sources/CodeCompass/parser/src/pluginhandler.cpp
> Segmentation fault (core dumped)

I have debugged that the segfault occurs in the `getMangledName()` function, here, when `fileLoc_.file.object_id()` gets evaluated:
```cpp
else if (llvm::isa<clang::NamespaceDecl>(nd_))
{
  const clang::NamespaceDecl* na_ = llvm::dyn_cast<clang::NamespaceDecl>(nd_);
  return na_->isAnonymousNamespace()
    ? "anonymous-ns:" + std::to_string(fileLoc_.file.object_id())
    : nd_->getQualifiedNameAsString();
}
```
https://codecompass.net/demo/#wsid=CodeCompass&fid=5704517689050331186&select=117|1|118|1&center=text

This can be followed to ODB's `lazy_ptr` implementation, where `i_.template object_id<O> ()` yields the segfault:
```cpp
template <class T>
template <class O>
inline typename object_traits<O>::id_type lazy_shared_ptr<T>::
object_id () const
{
  typedef typename object_traits<T>::object_type object_type;

  return p_
    ? object_traits<object_type>::id (*p_)
    : i_.template object_id<O> ();
}
```
https://codecompass.net/demo/#wsid=CodeCompass&fid=12316403250594715644&select=1351|1|1352|1&center=text

Which finally ends up here (with `r` being the culprit):
```cpp
template <typename T>
template <typename O>
inline typename object_traits<O>::id_type lazy_ptr_impl<T>::
object_id () const
{
  typedef typename object_traits<T>::id_type id_type;
  const id_type& id (*static_cast<const id_type*> (id_));

  // Make sure that O' and T's object id types are the same
  // (or implicit-convertible). If you get a compile error
  // pointing here, then you most likely used a wrong type
  // as a template argument in the object_id() call.
  //
  const typename object_traits<O>::id_type& r (id);
  return r;
}
```
https://codecompass.net/demo/#wsid=CodeCompass&fid=9722353701976068959&select=395|1|397|1&center=text

I tried and can parse some other projects, e.g. TinyXML and the first 23 TU's were parsed correctly even for CodeCompass.

**Comment by intjftw** [comment]
I managed to reproduce this error for the same file:

> [INFO] (21/138) Parsing /home/mate/cc/sources/CodeCompass/parser/src/pluginhandler.cpp
> Segmentation fault (core dumped)

Clean build of Ericsson master, 1 core.

**Comment by mcserep** [comment]
The bug is seemed to be caused by #425 .

**Comment by bruntib** [comment]
Segmentation fault happens in `getMangledName()` function which has been eliminated in #429. After merging that PR, this error will be fixed.

**Comment by mcserep** [comment]
> Segmentation fault happens in `getMangledName()` function which has been eliminated in #429. After merging that PR, this error will be fixed.

#429 also segfaults when trying to parse CodeCompass, see the CI pipeline for the PR.
Ofc the reason behind can be unrelated.

**Comment by mcserep** [comment]
Probably we shall revert #425, until this issue gets fixed, because the `master` branch is basically unusable for almost 6 weeks for now and it was not solved in almost 3 weeks after its first report.

**Comment by bruntib** [comment]
I'm pretty sure that the current Travis failure is a different issue, because the null pointer indirection which caused segmentation fault is not in the source code at all in #429. 
I couldn't reproduce this new failure on my local machine, so fixing that will require some more investigation.

**Comment by mcserep** [comment]
Fixed by switching to Usr (#429), so closing this.
